fix queries from direct space state(support for kinematic bodies). Also update to godot-cpp 4.1

Ughuuu opened the pull request:
Added implementation for:

fix _intersect_ray
fix _intersect_point
_intersect_shape
_cast_motion
_collide_shape
_rest_info

- Fixes https://github.com/rburing/physics_server_box2d/issues/12
- Fixes https://github.com/rburing/physics_server_box2d/issues/7
- Fixes https://github.com/rburing/physics_server_box2d/issues/11

rburing commented:
I think a point should not have a radius, no matter how small. See https://gamedev.stackexchange.com/a/1516/2351 which suggests to use `b2Fixture::TestPoint` in a loop over fixtures, after the AABB query.

Ughuuu commented:
Hm, I did something similar, but why would I not just use fixture->GetShape() ? Why get all the shapes of the body, if I know which shape collided.
Testing with the physics test I now also pass the 1 pixel away test.(Eg testing a point 1 pixel further than a shape and it doesnt return collision).

I think each shape has an AABB so it collides with each shape that might intersect, so that callback will be called at some point for all fixtures anyway, so there is no need to do foreach in there. It might be a performance issue if I do.
Will test a bit more after, rn doing implementation for the other functions also.

Ughuuu committed:
use godot branch 4.1

update to godot 4.1

update for 4.1

update submodules?

lint

update readme

update readme

add canvas instance id

Ughuuu commented:
In this PR I also updated to godot 4.1

Ughuuu committed:
add basic impl for shape collide

lint

lint

implement sweep test

add missing get_contact_local_velocity_at_position function for 4.1

write basic implementation for shape test

small error check

write correct implementation for _shape_collide

fix queries

lint

fix point intersect precision

fix constant central force

lint

remove unused

lint

write closest values

update readme

Ughuuu commented:
Added a basic implementation for these:

_intersect_shape
_cast_motion
_collide_shape
_rest_info

and for the one on physics_server:

_shape_collide

For the non motion vector ones:
_intersect_shape
_collide_shape
_collide_shape
_rest_info

I did a aabb check and then checked distance between all of them(not sure if there is a better solution)

For the ones that I think require motion check:
_cast_motion
_shape_collide

I did a aabb test(for ones that dont have 2 shapes only as param) and then a sweep test.

Ughuuu commented:
Ok, after testing it locally, seems to work. I am now passing 142/150 tests.

Ughuuu commented:
For point query, it's a simple aabb query and then test point containment, as you described.
For shape query, its a aabb query that returns some shapes(based on the aabb of the shapeA received as input) and then test distance between all resulting shapes and shapeA.
For motion query, it uses aabb query on begin/end location(to get all the shapes that might intersect with it), and then does a sweep query between shapeA and each shape.

Also updated the godot and box2d submodules, not sure if i did that correctly, let me know if it looks ok.

Ughuuu committed:
add basic impl for body_collide and body_motion

update those

exclude bodies in query

fix intersect point to pass all tests

update shape api

fix ray tests

fix raycast tests

Ughuuu committed:
update collision functions

update intersect shape(almost fixed, 2 more tests to fix)

fix direct space functions

fix max contacts count

upd

fix type conversion to match current godot space

fix density and space transform

refactor

fix pointer crash

add manifold code

update collision normal

fix intersect_shape tests

add shape collide impl

add impl for safe/unsafe fractions

apply b2Common hack

fix macos issue

update path of lib to addons/physics_server_box2d/

fix typo

rename artifactts

use wildcard pattern for addons folder

fix normal being zero

Ughuuu commented:
Updated the `intersects_shape` and other functions from direct space similar to it. Still something wrong with it when rotation objects. Also, if the shape starts already colliding with an object, it should not report collision(for test_move for eg.). That is also not yet done. Will open an issue for that also.

Started work on physics_server side for `_body_test_motion`, `_body_collide_shape` and `_shape_collide`.

Currently passing 107/150. 

Once those are also done it will be ready for review.

Ughuuu commented:
Updated the B_TO_G_FACTOR to be 100(since in godot gravity is 980, and in box2d its 9.8, means the factor has to be 100).

Also fix how density works and how inertia is computed for shapes(now instead of density of 1, density is put to be 1/computed_mass_of_shape). This way, the computed inertia is without the computed mass(box2d computes the object estimated mass, but we want to always hardcode it to a value, as is in godot). It also matches what godot does.

Currently passing 110/150 tests (a big part of these are CharacterBody2DTests, which need the correct impl for safe/unsafe fraction), as described here: https://github.com/rburing/physics_server_box2d/issues/18

rburing commented on src/shapes/box2d_shape.h:
```diff
@@ -15,14 +15,19 @@ class Box2DCollisionObject;
 
 class Box2DShape {
 	RID self;
+	Box2DCollisionObject *body;
```
I do not understand this. The `Box2DShape` is supposed to be a resource (hence the `RID`) that can be shared by multiple bodies. For example it should be possible to create 10 bodies all referring to one and the same circle shape resource. Can you test an example like that? The rest of the PR looks very good, but I don't think this part can work like this.

Ughuuu replied:
Hm, that would be a problem indeed. I guess it would have a reference to only one of the bodies.
Right now this is used to keep a track of the body, and it's used in case of eg. `_intersect_shape` to return collider_id and collider.
However, I see now that there is a method on the b2Fixture called GetBody, so I wouldn't need this reference, as I could use that.
Good catch.

rburing replied:
@Ughuuu indeed, that sounds like the correct approach.

rburing commented on src/shapes/box2d_shape.h:
```diff
@@ -31,7 +36,9 @@ class Box2DShape {
 	virtual Variant get_data() const = 0;
 
 	virtual int get_b2Shape_count(bool is_static) const = 0;
-	virtual b2Shape *get_transformed_b2Shape(int p_index, const Transform2D &p_transform, bool one_way, bool is_static) = 0;
```
The transform also isn't part of the resource (each body can transform the shape differently), which explains why I had the method like this.

Ughuuu replied:
Perfect, wasn't sure of that. Updating that.
